// angular2-grid/esm5/components/NgGridPlaceholder.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, Renderer } from '@angular/core';
var NgGridPlaceholder = /** @class */ (function () {
    function NgGridPlaceholder(_ngEl, _renderer) {
        this._ngEl = _ngEl;
        this._renderer = _renderer;
    }
    /**
     * @param {?} ngGrid
     * @return {?}
     */
    NgGridPlaceholder.prototype.registerGrid = /**
     * @param {?} ngGrid
     * @return {?}
     */
    function (ngGrid) {
        this._ngGrid = ngGrid;
    };
    /**
     * @return {?}
     */
    NgGridPlaceholder.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this._renderer.setElementClass(this._ngEl.nativeElement, 'grid-placeholder', true);
        if (this._ngGrid.autoStyle)
            this._renderer.setElementStyle(this._ngEl.nativeElement, 'position', 'absolute');
    };
    /**
     * @param {?} newSize
     * @return {?}
     */
    NgGridPlaceholder.prototype.setSize = /**
     * @param {?} newSize
     * @return {?}
     */
    function (newSize) {
        this._size = newSize;
        this._recalculateDimensions();
    };
    /**
     * @param {?} newPosition
     * @return {?}
     */
    NgGridPlaceholder.prototype.setGridPosition = /**
     * @param {?} newPosition
     * @return {?}
     */
    function (newPosition) {
        this._position = newPosition;
        this._recalculatePosition();
    };
    /**
     * @param {?} cascade
     * @return {?}
     */
    NgGridPlaceholder.prototype.setCascadeMode = /**
     * @param {?} cascade
     * @return {?}
     */
    function (cascade) {
        this._cascadeMode = cascade;
        switch (cascade) {
            case 'up':
            case 'left':
            default:
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'left', '0px');
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'top', '0px');
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'right', null);
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'bottom', null);
                break;
            case 'right':
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'right', '0px');
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'top', '0px');
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'left', null);
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'bottom', null);
                break;
            case 'down':
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'left', '0px');
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'bottom', '0px');
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'right', null);
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'top', null);
                break;
        }
    };
    /**
     * @param {?} w
     * @param {?} h
     * @return {?}
     */
    NgGridPlaceholder.prototype._setDimensions = /**
     * @param {?} w
     * @param {?} h
     * @return {?}
     */
    function (w, h) {
        this._renderer.setElementStyle(this._ngEl.nativeElement, 'width', w + 'px');
        this._renderer.setElementStyle(this._ngEl.nativeElement, 'height', h + 'px');
    };
    /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    NgGridPlaceholder.prototype._setPosition = /**
     * @param {?} x
     * @param {?} y
     * @return {?}
     */
    function (x, y) {
        switch (this._cascadeMode) {
            case 'up':
            case 'left':
            default:
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'transform', 'translate(' + x + 'px, ' + y + 'px)');
                break;
            case 'right':
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'transform', 'translate(' + -x + 'px, ' + y + 'px)');
                break;
            case 'down':
                this._renderer.setElementStyle(this._ngEl.nativeElement, 'transform', 'translate(' + x + 'px, ' + -y + 'px)');
                break;
        }
    };
    /**
     * @return {?}
     */
    NgGridPlaceholder.prototype._recalculatePosition = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ x = (this._ngGrid.colWidth + this._ngGrid.marginLeft + this._ngGrid.marginRight) * (this._position.col - 1) + this._ngGrid.marginLeft + this._ngGrid.screenMargin;
        var /** @type {?} */ y = (this._ngGrid.rowHeight + this._ngGrid.marginTop + this._ngGrid.marginBottom) * (this._position.row - 1) + this._ngGrid.marginTop;
        this._setPosition(x, y);
    };
    /**
     * @return {?}
     */
    NgGridPlaceholder.prototype._recalculateDimensions = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ w = (this._ngGrid.colWidth * this._size.x) + ((this._ngGrid.marginLeft + this._ngGrid.marginRight) * (this._size.x - 1));
        var /** @type {?} */ h = (this._ngGrid.rowHeight * this._size.y) + ((this._ngGrid.marginTop + this._ngGrid.marginBottom) * (this._size.y - 1));
        this._setDimensions(w, h);
    };
    NgGridPlaceholder.decorators = [
        { type: Component, args: [{
                    selector: 'ng-grid-placeholder',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    NgGridPlaceholder.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer }
    ]; };
    return NgGridPlaceholder;
}());
export { NgGridPlaceholder };
function NgGridPlaceholder_tsickle_Closure_declarations() {
    /** @type {?} */
    NgGridPlaceholder.prototype._size;
    /** @type {?} */
    NgGridPlaceholder.prototype._position;
    /** @type {?} */
    NgGridPlaceholder.prototype._ngGrid;
    /** @type {?} */
    NgGridPlaceholder.prototype._cascadeMode;
    /** @type {?} */
    NgGridPlaceholder.prototype._ngEl;
    /** @type {?} */
    NgGridPlaceholder.prototype._renderer;
}
