// angular2-grid/esm5/interfaces/INgGrid.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function NgGridConfig() { }
function NgGridConfig_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    NgGridConfig.prototype.margins;
    /** @type {?|undefined} */
    NgGridConfig.prototype.draggable;
    /** @type {?|undefined} */
    NgGridConfig.prototype.resizable;
    /** @type {?|undefined} */
    NgGridConfig.prototype.max_cols;
    /** @type {?|undefined} */
    NgGridConfig.prototype.max_rows;
    /** @type {?|undefined} */
    NgGridConfig.prototype.visible_cols;
    /** @type {?|undefined} */
    NgGridConfig.prototype.visible_rows;
    /** @type {?|undefined} */
    NgGridConfig.prototype.min_cols;
    /** @type {?|undefined} */
    NgGridConfig.prototype.min_rows;
    /** @type {?|undefined} */
    NgGridConfig.prototype.col_width;
    /** @type {?|undefined} */
    NgGridConfig.prototype.row_height;
    /** @type {?|undefined} */
    NgGridConfig.prototype.cascade;
    /** @type {?|undefined} */
    NgGridConfig.prototype.min_width;
    /** @type {?|undefined} */
    NgGridConfig.prototype.min_height;
    /** @type {?|undefined} */
    NgGridConfig.prototype.fix_to_grid;
    /** @type {?|undefined} */
    NgGridConfig.prototype.auto_style;
    /** @type {?|undefined} */
    NgGridConfig.prototype.auto_resize;
    /** @type {?|undefined} */
    NgGridConfig.prototype.maintain_ratio;
    /** @type {?|undefined} */
    NgGridConfig.prototype.prefer_new;
    /** @type {?|undefined} */
    NgGridConfig.prototype.zoom_on_drag;
    /** @type {?|undefined} */
    NgGridConfig.prototype.limit_to_screen;
    /** @type {?|undefined} */
    NgGridConfig.prototype.center_to_screen;
    /** @type {?|undefined} */
    NgGridConfig.prototype.resize_directions;
    /** @type {?|undefined} */
    NgGridConfig.prototype.element_based_row_height;
    /** @type {?|undefined} */
    NgGridConfig.prototype.fix_item_position_direction;
    /** @type {?|undefined} */
    NgGridConfig.prototype.fix_collision_position_direction;
    /** @type {?|undefined} */
    NgGridConfig.prototype.allow_overlap;
}
/**
 * @record
 */
export function NgGridItemConfig() { }
function NgGridItemConfig_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.uid;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.payload;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.col;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.row;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.sizex;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.sizey;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.dragHandle;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.resizeHandle;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.fixed;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.draggable;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.resizable;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.borderSize;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.maxCols;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.minCols;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.maxRows;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.minRows;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.minWidth;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.minHeight;
    /** @type {?|undefined} */
    NgGridItemConfig.prototype.resizeDirections;
}
/**
 * @record
 */
export function NgGridItemEvent() { }
function NgGridItemEvent_tsickle_Closure_declarations() {
    /** @type {?} */
    NgGridItemEvent.prototype.uid;
    /** @type {?} */
    NgGridItemEvent.prototype.payload;
    /** @type {?} */
    NgGridItemEvent.prototype.col;
    /** @type {?} */
    NgGridItemEvent.prototype.row;
    /** @type {?} */
    NgGridItemEvent.prototype.sizex;
    /** @type {?} */
    NgGridItemEvent.prototype.sizey;
    /** @type {?} */
    NgGridItemEvent.prototype.width;
    /** @type {?} */
    NgGridItemEvent.prototype.height;
    /** @type {?} */
    NgGridItemEvent.prototype.left;
    /** @type {?} */
    NgGridItemEvent.prototype.top;
}
/**
 * @record
 */
export function NgGridItemSize() { }
function NgGridItemSize_tsickle_Closure_declarations() {
    /** @type {?} */
    NgGridItemSize.prototype.x;
    /** @type {?} */
    NgGridItemSize.prototype.y;
}
/**
 * @record
 */
export function NgGridItemPosition() { }
function NgGridItemPosition_tsickle_Closure_declarations() {
    /** @type {?} */
    NgGridItemPosition.prototype.col;
    /** @type {?} */
    NgGridItemPosition.prototype.row;
}
/**
 * @record
 */
export function NgGridRawPosition() { }
function NgGridRawPosition_tsickle_Closure_declarations() {
    /** @type {?} */
    NgGridRawPosition.prototype.left;
    /** @type {?} */
    NgGridRawPosition.prototype.top;
}
/**
 * @record
 */
export function NgGridItemDimensions() { }
function NgGridItemDimensions_tsickle_Closure_declarations() {
    /** @type {?} */
    NgGridItemDimensions.prototype.width;
    /** @type {?} */
    NgGridItemDimensions.prototype.height;
}
